File: 173-binary-search-tree-iterator/binary-search-tree-iterator.java
public class BSTIterator {
    private Stack<TreeNode> stack = new Stack<>();

    public BSTIterator(TreeNode node) {
        this.addAllLeftChilds(node);
    }

    private void addAllLeftChilds(TreeNode root) {
        while (root != null) {
            this.stack.add(root);
            // move left side only
            root = root.left;
        }
    }

    public int next() {
        // if hasnext
        // then pop the elem from stack
        // check if it has right child
        // add all the right.leftchilds of the subtree into stack
        if (this.hasNext()) {
            TreeNode node = this.stack.pop();
            // has right?
            if (node.right != null) {
                this.addAllLeftChilds(node.right);
            }

            return node.val;
        }
        return -1;
    }

    public boolean hasNext() {
        // return if stack is empty
        return !this.stack.isEmpty();
    }

}

File: 173-binary-search-tree-iterator/binary-search-tree-iterator.py
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.stack=[]
        self.add_all_left(root)

    def add_all_left(self, root):
        while root:
            self.stack.append(root)
            root=root.left

    def next(self) -> int:
        if self.hasNext():
            node=self.stack.pop()
            # does it has right
            if node.right is not None:
                self.add_all_left(node.right)

            return node.val
        
        return -1

    def hasNext(self) -> bool:
        return len(self.stack)>0
        